src/mesh/gmsh_io.C

```diff
@@ -275,10 +283,96 @@ void GmshIO::read_mesh(std::istream & in)
                 }
             }
 
+          else if (s.find("$Entities") == static_cast<std::string::size_type>(0))
+          {
+            if (version >= 4.0)
+            {
+              size_t num_point_entities, num_curve_entities, num_surface_entities, num_volume_entities;
+              in >> num_point_entities >> num_curve_entities >> num_surface_entities >> num_volume_entities;
+
+              for (size_t n = 0; n < num_point_entities; ++n)
+              {
+                int point_tag, physical_tag;
+                Real x, y, z;
+                size_t num_physical_tags;
+                in >> point_tag >> x >> y >> z >> num_physical_tags;
+                if (num_physical_tags > 1)
+                  libmesh_error_msg("I don't believe that we can specify multiple subdomain or " <<
+                                    "boundary ids for a given geometric entity");
+                else if (num_physical_tags)
+                {
+                  in >> physical_tag;
+                  entity_to_physical_id[std::make_pair(0, point_tag)] = physical_tag;
+                }
+              }
+              for (size_t n = 0; n < num_curve_entities; ++n)
+              {
+                int curve_tag, physical_tag;
+                Real minx, miny, minz, maxx, maxy, maxz;
+                size_t num_physical_tags;
+                in >> curve_tag >> minx >> miny >> minz >> maxx >> maxy >> maxz >> num_physical_tags;
+                if (num_physical_tags > 1)
+                  libmesh_error_msg("I don't believe that we can specify multiple subdomain or " <<
+                                    "boundary ids for a given geometric entity");
+                else if (num_physical_tags)
+                {
+                  in >> physical_tag;
+                  entity_to_physical_id[std::make_pair(1, curve_tag)] = physical_tag;
+                }
+
+                // Read to end of line; this captures bounding information that we don't care about
+                std::getline(in, s);
+              }
+              for (size_t n = 0; n < num_surface_entities; ++n)
```

Comment:
We should try to be consistent and always use `std::size_t`. We haven't always done this in the past, but I'm trying to be better about it. We should always be #includeing `<cstddef>` which puts that type into the std namespace, and some compilers are pickier about this than others...